frontend/catalyst/cuda/catalyst_to_cuda_interpreter.py

```diff
@@ -0,0 +1,916 @@
+        """Keep track of which wire this qubit variable is."""
+        self.qubit_to_wire_map[qubit] = idx
+
+    def get_wire_for_qubit(self, qubit):
+        """Get the wire for this qubit."""
+        return self.qubit_to_wire_map[qubit]
+
+    def read(self, var):
+        """Read the value of variable var."""
+        if isinstance(var, jax.core.Literal):
+            return var.val
+        if self.variable_map.get(var):
+            var = self.variable_map[var]
+        return self.env[var]
+
+    def write(self, var, val):
+        """var = val."""
+        if self.variable_map.get(var):
+            var = self.variable_map[var]
+        self.env[var] = val
+
+    def replace(self, original, new):
+        """Replace original variable with new variable."""
+        self.variable_map[original] = new
+
+    def get_new_count(self):
+        """Increase count and return."""
+        self.count += 1
+        return self.count
+
+    def new_variable(self):
+        """Convenience to get a new variable of a given type.
+
+        These are not JAX variables. This can be any hashable.
+        We are using integers as a small level of indirection.
+        Each new variable is a new integer.
+
+        This is pretty much what JAX does since each variable gets
+        assigned a count and a type. But it is also what anyone else does.
+        (Like LLVM/MLIR %0, %1, ...)
+
+        TODO(@erick-xanadu): I think the type information
+        would also be good to have here, but it is unnecessary at the
+        moment.
+        """
+        return self.get_new_count()
+
+
+def change_device_to_cuda_device(ctx):
+    """Map Catalyst's qdevice_p primitive to its equivalent CUDA-quantum primitive
+    as defined in this file.
+
+    From here we get shots as well.
+    """
+
+    # The device here might also have some important information for
+    # us. For example, the number of shots.
+
+    qdevice_eqn = get_instruction(ctx.jaxpr, qdevice_p)
+
+    # These parameters are stored in a json-like string.
+    # We first convert the string to json.
+    json_like_string = qdevice_eqn.params["rtd_kwargs"]
+    json_like_string = json_like_string.replace("'", '"')
+    json_like_string = json_like_string.replace("True", "true")
+    json_string = json_like_string.replace("False", "false")
+
+    # Finally, we load it
+    parameters = json.loads(json_string)
+
+    # Now we have the number of shots.
+    # Shots are specified in PL at the very beginning, but in cuda
+    # shots are not needed until the very end.
+    # So, we will just return this variable
+    # and it is the responsibility of the caller to propagate this information.
+    shots = parameters.get("shots")
+
+    device_name = qdevice_eqn.params.get("rtd_name")
+
+    # TODO(@erick-xanadu) as more devices become available
+    # map the names here.
+    # TODO(@erick-xanadu) why does the device instruction lists the whole
+    # name instead of a short name?
```

Comment:
I added this ticket, which is relevant: https://github.com/PennyLaneAI/catalyst/issues/517 And also relevant to the discussion above about `qdevice_p`.

Reply:
Well the `qdevice_p` primitive is there to enable the device to function. For Catalyst devices it needs the name of the C++ device so it can find the right entry point function in the low-level library (it is independent of whatever long or short name the Python device class has). For this compilation pipeline we are free to supply whatever name is suitable for it.

Maybe this can solve the issue in the discussion above by getting rid of the map entirely? Instead the `qdevice_p` primitive is already supplied by the right backend name upon instantiation, which can be sourced from the Python device class (similar to how `get_c_interface` obtains the right device name in Catalyst).

Reply:
> Maybe this can solve the issue in the discussion above by getting rid of the map entirely? Instead the qdevice_p primitive is already supplied by the right backend name upon instantiation, which can be sourced from the Python device class (similar to how get_c_interface obtains the right device name in Catalyst).

I see what you mean. Thanks!